llrt_utils/src/bytes.rs

```diff
@@ -1,8 +1,247 @@
+    const DETACHED_ERROR: &'static str = "ArrayBuffer is detached";
+
+    pub fn from(ctx: &Ctx<'js>, value: Value<'js>) -> Result<Self> {
+        Self::from_offset(ctx, value, 0, None)
+    }
+
+    pub fn from_offset(
+        ctx: &Ctx<'js>,
+        value: Value<'js>,
+        offset: usize,
+        length: Option<usize>,
+    ) -> Result<Self> {
+        if value.is_undefined() {
+            return Ok(ObjectBytes::Vec(Some(vec![])));
+        }
+        if let Some(bytes) = get_string_bytes(&value, offset, length)? {
+            return Ok(ObjectBytes::Vec(Some(bytes)));
+        }
+        if let Some(bytes) = get_array_bytes(&value, offset, length)? {
+            return Ok(ObjectBytes::Vec(Some(bytes)));
+        }
+
+        if let Some(obj) = value.as_object() {
+            if let Some(bytes) = Self::from_array_buffer(obj)? {
+                return Ok(bytes);
+            }
+        }
+
+        if let Some(bytes) = get_coerced_string_bytes(&value, offset, length) {
+            return Ok(ObjectBytes::Vec(Some(bytes)));
+        }
+
+        Err(Exception::throw_message(
+        ctx,
+        "value must be typed DataView, Buffer, ArrayBuffer, Uint8Array or interpretable as string",
+    ))
+    }
+
+    pub fn get_bytes(&mut self) -> Cow<'_, [u8]> {
+        let cow = match self {
+            ObjectBytes::U8Array(array) => {
+                Cow::Borrowed(array.as_bytes().expect(Self::DETACHED_ERROR))
+            },
+            ObjectBytes::I8Array(array) => {
+                Cow::Borrowed(array.as_bytes().expect(Self::DETACHED_ERROR))
+            },
+            ObjectBytes::U16Array(array) => {
+                Cow::Borrowed(array.as_bytes().expect(Self::DETACHED_ERROR))
+            },
+            ObjectBytes::I16Array(array) => {
+                Cow::Borrowed(array.as_bytes().expect(Self::DETACHED_ERROR))
+            },
+            ObjectBytes::U32Array(array) => {
+                Cow::Borrowed(array.as_bytes().expect(Self::DETACHED_ERROR))
+            },
+            ObjectBytes::I32Array(array) => {
+                Cow::Borrowed(array.as_bytes().expect(Self::DETACHED_ERROR))
+            },
+            ObjectBytes::U64Array(array) => {
+                Cow::Borrowed(array.as_bytes().expect(Self::DETACHED_ERROR))
+            },
+            ObjectBytes::I64Array(array) => {
+                Cow::Borrowed(array.as_bytes().expect(Self::DETACHED_ERROR))
+            },
+            ObjectBytes::F32Array(array) => {
+                Cow::Borrowed(array.as_bytes().expect(Self::DETACHED_ERROR))
+            },
+            ObjectBytes::F64Array(array) => {
+                Cow::Borrowed(array.as_bytes().expect(Self::DETACHED_ERROR))
+            },
+            ObjectBytes::DataView(array_buffer) => {
+                Cow::Borrowed(array_buffer.as_bytes().expect(Self::DETACHED_ERROR))
+            },
+            ObjectBytes::Vec(bytes) => Cow::Owned(bytes.take().expect("Bytes already taken")),
```

Comment:
I feel like we could always borrow in this method and get rid of the cow. Currently it is a mix of both which will be confusing since there is no guarantee that a second call will return data.

Reply:
It’s for performance reasons. If we only borrow we need an additional clone to get get the underlying Vec. Agree it’s a bit strange that calling get_bytes() a second time would panic if already called but I couldn’t think of another way. Suggestions are welcome!

Reply:
Yes I see we have two use cases. So lets do two functions.
- When you are passing the return value has a reference to say `.write` use the `as_bytes` method which takes `&self` and returns a slice.
- When you are using an owned version say in the hasher, use `.into_bytes` which consumes `self` and returns a `vec`.